File: ICorteApi/Application/Validators/SharingValidators/PasswordValidator.cs
using FluentValidation;

namespace ICorteApi.Application.Validators;

internal static class PasswordValidationExtensions
{
    public static IRuleBuilderOptions<T, string> ApplyPasswordValidation<T>(
        this IRuleBuilder<T, string> ruleBuilder,
        string? messageIdentifier = null)
    {
        string Senha = messageIdentifier ?? "Senha";
        
        return ruleBuilder
            .NotEmpty().WithMessage($"{Senha} obrigatória")
            .DependentRules(() =>
            {
                ruleBuilder
                    .MinimumLength(8).WithMessage($"{Senha} deve conter pelo menos 8 caracteres")
                    .Must(ContainUppercase).WithMessage($"{Senha} deve conter pelo menos uma letra maiúscula")
                    .Must(ContainLowercase).WithMessage($"{Senha} deve conter pelo menos uma letra minúscula")
                    .Must(ContainDigit).WithMessage($"{Senha} deve conter pelo menos um dígito")
                    .Must(ContainNonAlphanumeric).WithMessage($"{Senha} deve conter pelo menos um caractere especial");
            });
    }

    private static bool IsNull(string? value) => value is null;

    private static bool ContainUppercase(string? password) => !IsNull(password) && password!.Any(char.IsUpper);
    private static bool ContainLowercase(string? password) => !IsNull(password) && password!.Any(char.IsLower);
    private static bool ContainDigit(string? password) => !IsNull(password) && password!.Any(char.IsDigit);
    private static bool ContainNonAlphanumeric(string? password) => !IsNull(password) && password!.Any(ch => !char.IsLetterOrDigit(ch));
}

File: client-app/src/schemas/sharedValidators/passwordValidator.ts
import { z } from 'zod'

const cotainsUpperCase = (value: string) => /[A-Z]/.test(value)
const cotainsLowerCase = (value: string) => /[a-z]/.test(value)
const cotainsDigit = (value: string) => /[0-9]/.test(value)
const cotainsNonAlphanumeric = (value: string) => /[!@#$%^&*(),.?":{}|<>]/.test(value)

export function passwordValidator(messageIdentifier?: string) {
  const Senha = messageIdentifier ?? 'Senha'

  return z.string()
    .min(1, { message: `${Senha} obrigatória` })
    .min(8, { message: `${Senha} deve conter pelo menos 8 caracteres` })
    .refine(cotainsUpperCase, { message: `${Senha} deve conter pelo menos uma letra maiúscula` })
    .refine(cotainsLowerCase, { message: `${Senha} deve conter pelo menos uma letra minúscula` })
    .refine(cotainsDigit, { message: `${Senha} deve conter pelo menos um dígito` })
    .refine(cotainsNonAlphanumeric, { message: `${Senha} deve conter pelo menos um caractere especial` })
}
